keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java

```diff
@@ -395,22 +397,25 @@ public DualNumber reshape(int[] proposedShape) {
     }
 
     public DualNumber slice(int dimension, int index) {
-        PartialDerivatives slicedPartialDerivatives = this.partialDerivatives.slice(dimension, index);
-        return new DualNumber(value.slice(dimension, index), slicedPartialDerivatives);
+        DoubleTensor newValue = value.slice(dimension, index);
+        boolean needReshape = newValue.getRank() == value.getRank();
+        PartialDerivatives slicedPartialDerivatives = this.partialDerivatives.slice(dimension, index, needReshape);
+        return new DualNumber(newValue, slicedPartialDerivatives);
     }
 
     public DualNumber take(int... index) {
         Map<VertexId, DoubleTensor> dualsAtIndex = new HashMap<>();
+        DoubleTensor newValue = DoubleTensor.scalar(this.value.getValue(index));
 
         for (Map.Entry<VertexId, DoubleTensor> entry : this.partialDerivatives.asMap().entrySet()) {
             DoubleTensor atIndexTensor = takeFromPartial(entry.getValue(), index);
-            int desiredRank = entry.getValue().getShape().length;
+            int desiredRank = atIndexTensor.getShape().length + newValue.getShape().length;
```

Comment:
Tensor.SCALAR_SHAPE.length is good but then it's like why scalar? And that's because it's the newValue shape, which you're right is always a scalar in this case. Because Tensor.SCALAR_SHAPE.length is always the shape that DoubleTensor.scalar returns this would work but I think it hides why and is therefore less clear.